### ISSUE TITLE
[Performance] JIT cache lookup costs ~1.8 s per process; target under 1 s

### ISSUE THREAD

**Issue by lyfne123** [hw-native-sys/pypto#2881]
### Summary

A persistent JIT cache hit costs about 1.8 s per process before any kernel is
served, and that cost is paid again by every new process. The work is the
installation identity capture in `pypto.jit._toolchain.capture_toolchain`, which
is memoized per process and so never amortizes across processes.

The target for this issue is to bring the whole per-process lookup under 1 s.

This issue records only what is measured today. It does not propose a design.
It supersedes #2732, which framed the same cost around one particular remedy;
the measurements below are the part worth keeping.

Related: #2653

### Git Commit ID

0937739db20bdb858e56e641fdd75a4699bbfcfa

### NPU Kind

N/A (CPU-only issue)

### Host Platform

Linux (aarch64)

### Reproduction Script

```python
"""Measure the per-process cost of a persistent JIT cache hit.

Run twice against the same cache root: the first run populates it, the second
measures a warm cross-process hit. No NPU is initialised.

    python repro.py /tmp/pypto-cache   # cold: builds and publishes
    python repro.py /tmp/pypto-cache   # warm: every kernel is a READY hit
"""

import sys
import time
from pathlib import Path

import pypto
import pypto.language as pl
import pypto.jit._toolchain as _toolchain
import torch
from pypto.runtime import RunConfig


@pl.jit
def tile_add(a: pl.Tensor, b: pl.Tensor, c: pl.Out[pl.Tensor]):
    with pl.at(level=pl.Level.CORE_GROUP):
        pl.store(pl.add(pl.load(a, [0, 0], [128, 128]), pl.load(b, [0, 0], [128, 128])), [0, 0], c)
    return c


@pl.jit
def tile_mul(a: pl.Tensor, b: pl.Tensor, c: pl.Out[pl.Tensor]):
    with pl.at(level=pl.Level.CORE_GROUP):
        pl.store(pl.mul(pl.load(a, [0, 0], [128, 128]), pl.load(b, [0, 0], [128, 128])), [0, 0], c)
    return c


def main() -> None:
    root = Path(sys.argv[1] if len(sys.argv) > 1 else "pypto-cache")
    cfg = RunConfig(platform="a2a3", cache_config=pypto.CacheConfig(enabled=True, root=root))
    x = torch.zeros(128, 128)

    # Toolchain identity is memoised per process, so time it on its own first.
    t0 = time.perf_counter()
    identity = _toolchain.capture_toolchain("a2a3", "tensormap_and_ringbuffer")
    t_identity = time.perf_counter() - t0
    assert identity.usable, identity.failures

    per_kernel = []
    for fn in (tile_add, tile_mul):
        t0 = time.perf_counter()
        fn.warmup(x, x, x, config=cfg)
        per_kernel.append((fn.__name__, time.perf_counter() - t0))

    stats = pypto.cache_stats()
    print(f"cache root          : {root}")
    print(f"toolchain identity  : {t_identity:.3f} s   (once per process)")
    for name, dt in per_kernel:
        print(f"{name:<20}: {dt:.3f} s")
    print(f"ready_hits={stats.ready_hits} generated_hits={stats.generated_hits} "
          f"misses={stats.misses} bypasses={stats.bypasses}")
    print(f"lookup_ns={stats.lookup_ns / 1e9:.3f} s  build_ns={stats.build_ns / 1e9:.3f} s")
    if stats.last_bypass_reason:
        print(f"last_bypass_reason  : {stats.last_bypass_reason}")


if __name__ == "__main__":
    main()
```

### Expected Performance

Under 1 s for the complete per-process lookup on a warm cache, measured as the
`toolchain identity` line of the reproduction script plus the first kernel's
`warmup`.

### Actual Performance

Warm second run, `PYPTO_CACHE` enabled, all kernels `ready_hits`, no bypass:

```
toolchain identity  : 1.770 s   (once per process)
tile_add            : 0.129 s
tile_mul            : 0.013 s
ready_hits=2 generated_hits=0 misses=0 bypasses=0
lookup_ns=0.057 s  build_ns=0.000 s
```

The identity capture dominates: 1.77 s against 0.13 s for everything the first
kernel then needs, and 0.01 s for each kernel after it. Ten consecutive warm
runs gave a median identity of 1.76 s (min 1.65 s, max 1.79 s).

For scale, on the same host and commit a small kernel compiles from scratch in
about 1.9 s with the cache disabled. A process that serves only one small kernel
therefore gains nothing from a cache hit; the break-even is two kernels.

### Profiling Data (Optional)

All figures below are from the same host and commit, warm page cache, median of
three runs unless stated. `capture_toolchain` totalled 1.766 / 1.795 / 1.866 s
across the three.

Two phases:

| Phase | Time | Share |
| ----- | ---- | ----- |
| `_discover` (resolve the effective tools) | 0.65–0.72 s | 37% |
| `InstallationIdentityCache.capture` (hash the inventories) | 1.10–1.13 s | 62% |

Inside `capture`, over 8924 files / 370 MB actually hashed:

| Work | Time |
| ---- | ---- |
| `_file_digest` (ELF-filtered content hashing) | 0.65–0.67 s |
| directory traversal and entry construction | ~0.45 s |

Per component:

| Component | Roots | Time | Files | MB on disk |
| --------- | ----- | ---- | ----- | ---------- |
| `device_toolchain` | 11 | 0.565–0.582 s | 7042 | 94.8 |
| `pypto` | 238 | 0.285–0.312 s | 1230 | 457.0 |
| `runtime` | 81 | 0.247–0.252 s | 672 | 349.3 |
| `pto_isa` | 0 | 0.000 s | — | — |
| `ptoas` | 0 | 0.000 s | — | — |

`pto_isa` and `ptoas` cost nothing because they are identified by a verified
revision rather than by reading content (#2825, #2849).

`device_toolchain` is dominated by one root. `/usr/include` alone accounts for
0.474–0.488 s over 7098 entries — roughly 27% of the entire lookup. It enters
the inventory through `_include_roots` parsing `ccec -E -v`: the BiSheng device
compiler searches the host include path. Measured with `-M` on the real
generated sources and the real flags, a compile reaches 17 files under
`/usr/include` and `/usr/local/include`, out of 7033 present. The orchestration
compile reaches none of them, because CANN's `aarch64-target-linux-gnu-g++` is a
cross compiler with its own sysroot.

Note `pypto` holds 457 MB on disk but costs less than `device_toolchain`'s
94.8 MB: the cost tracks file count, not bytes. ELF debug-payload filtering
(#2862) already reduces 901 MB of candidate content to the 370 MB hashed.

Inside `_discover`:

| Work | Time | Calls |
| ---- | ---- | ----- |
| `ldd` dependency closures | 0.263 s | 109 subprocesses (99 pooled, 10 serial) |
| `_run` (`ccec -E -v`, `gcc -###`) | 0.134 s | 18 subprocesses |
| `_package` | 0.121 s | 3 |
| `_gcc_inputs` | 0.095 s | 1 |
| `_ptoas_component` | 0.082 s | 1 |

Traversal cost, measured by serving every digest from a prepared table so only
the walk remains (0.283 s over all roots):

| Cumulative | Time |
| ---------- | ---- |
| `os.walk` only | 0.018 s |
| + `scandir` / sort / `stat` | 0.039 s |
| + re-`scandir` and per-file `lstat` (replacement detection) | 0.072 s |
| + `Path` objects and entry tuples | 0.194 s |

Replacement detection is cheap at 0.033 s; the remainder is per-entry Python
object construction.

### Additional Context

Measured on a shared 320-core aarch64 host (openEuler, glibc 2.36, GCC 15.2.1,
CANN 9.0.0, PTOAS v0.65), Python 3.10, in a clean worktree built from the commit
above. Under heavy load from other tenants the same measurement reached 2.5–5.2 s,
so the figures here were taken from consecutive runs at a stable load and the
spread is reported alongside each.

Persistent caching is still opt-in at this commit (`PYPTO_CACHE` defaults to 0),
so this cost is paid only by callers that enable it. `docs/en/dev/10-jit-cache.md`
currently states "approximately 12 seconds" for a first content inventory; that
figure predates #2810, #2824, #2825, #2849 and #2862 and no longer reproduces.
